design/generic_resources.md

```diff
@@ -0,0 +1,158 @@
+    * GPU devices
+    * FPGA devices
+    * MICs (Many-Integrated Core, such as Xeon Phi)
+    * ...
+  * Support for tracking additional cgroup quotas like cpu_rt_runtime.
+    * [Linux Realtime](https://github.com/docker/docker/pull/23430)
+  * PersistentDisks in GCE
+  * Counting “slots” allowed access to a shared parallel file system.
+
+## Related Issues
+
+  * [Support abstract resource](https://github.com/docker/swarmkit/issues/594)
+  * [Add new node filter to scheduler](https://github.com/docker/swarm/issues/2223)
+  * [Add support for devices](https://github.com/docker/swarmkit/issues/1244)
+  * [Resource Control](https://github.com/docker/swarmkit/issues/211)
+  * [NVIDIA GPU support](https://github.com/docker/docker/issues/23917)
+  * [Does Docker have plan to support allocating GPU](https://github.com/docker/docker/issues/24582)
+  * [Docker Swarm to orchestrate "Swarm Cluster" which supports GPU](https://github.com/docker/docker/issues/24750)
+  * [Use accelerator in docker container](https://github.com/docker/docker/issues/28642)
+  * [Specify resource selectors](https://github.com/docker/swarmkit/issues/206)
+
+## Objectives
+
+1. Associate multiple generic resources with a node
+2. Request some portion of available generic resources in the service
+   during service creation
+3. Enable users to define and schedule generic resources in a vanilla swarmkit cluster
+
+## Non-Objectives
+
+1. Solve how generic resources allocations are to be enforced or isolated.
+2. Solve how generic resources are discovered
+2. Solve how to filter at the resources level
+3. Solve how cluster-level generic resources should be advertised
+
+## Proposed Changes
+
+### Generic Resources request
+
+The services may only ask for generic resources as integers as the solution for asking for
+specific resources can be solved in many different ways (filters, multiple kinds
+of resources, ...) and should not be addressed in this PR.
+
+```
+$ # Single resource
+$ swarmctl service create --name nginx --image nginx:latest --generic-resources "banana=2"
+$ # Multiple resources
+$ swarmctl service create --name nginx --image nginx:latest --generic-resources "banana=2;apple=3"
+```
+
+### Generic Resource advertising
+
+A node may advertise either an discrete number of resources or a set of resources.
+It is the scheduler's job to decide which resource to assign and keep track of which task
+owns which resource.
+
+```
+$ swarmd -d $DIR --join-addr $IP --join-token $TOKEN --generic-node-resources "banana={blue,red,green};apple=8"
+```
+
+### Generic Resource communication
+
+As swarmkit is not responsible for the allocation of resources, it needs a way
+to communicate how many generic resources were assigned (in the case of discrete resources)
+and what resources were selected (in the case of sets).
+
+The way we chose to implement this is through environment variables as they seem
+to be the most straightforward way to advertise to the running container this.
```

Comment:
Agreed, @RenaudWasTaken can you prefix it with `DOCKER_RESOURCE_` and add a `ToUpper` in there.